README.md
# formstream

[![NPM version][npm-image]][npm-url]
[![Test coverage][coveralls-image]][coveralls-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/formstream.svg?style=flat-square
[npm-url]: https://npmjs.org/package/formstream
[coveralls-image]: https://img.shields.io/coveralls/node-modules/formstream.svg?style=flat-square
[coveralls-url]: https://coveralls.io/r/node-modules/formstream?branch=master
[download-image]: https://img.shields.io/npm/dm/formstream.svg?style=flat-square
[download-url]: https://npmjs.org/package/formstream

A [multipart/form-data](http://tools.ietf.org/html/rfc2388) encoded stream, helper for file upload.

## Install

```bash
$ npm install formstream
```

## Quick Start

```js
var formstream = require('formstream');
var http = require('http');

var form = formstream();

// form.file('file', filepath, filename);
form.file('file', './logo.png', 'upload-logo.png');

// other form fields
form.field('foo', 'fengmk2').field('love', 'aerdeng');

// even send file content buffer directly
// form.buffer(name, buffer, filename, mimeType)
form.buffer('file2', new Buffer('This is file2 content.'), 'foo.txt');

var options = {
  method: 'POST',
  host: 'upload.cnodejs.net',
  path: '/store',
  headers: form.headers()
};
var req = http.request(options, function (res) {
  console.log('Status: %s', res.statusCode);
  res.on('data', function (data) {
    console.log(data.toString());
  });
});

form.pipe(req);
```

### Chaining

```js
var fs = require('fs');
var formstream = require('formstream');

var filepath = './logo.png';
fs.stat(filepath, function (err, stat) {
  formstream().field('status', 'share picture')
      .field('access_token', 'your access token')
      .file('pic', filepath, 'logo.png', stat.size)
      .pipe(process.stdout); // your request stream
});
```

## API Doc

### formstream()

Create a form instance.

#### Returns

Form - form instance

### FormStream#field(name, value)

Add a normal field to the form.

#### Arguments

- **name** String - Name of field
- **value** String - Value of field

#### Returns

Form - form instance

### FormStream#file(name, filepath[, filename][, filesize])

Add a local file to be uploaded to the form.

#### Arguments

- **name** String - Name of file field
- **filepath** String - Local path of the file to be uploaded
- ***filename*** String - Optional. Name of the file (will be the base name of `filepath` if empty)
- ***filesize*** Number - Optional. Size of the file (will not generate `Content-Length` header if not specified)

#### Returns

Form - form instance

### FormStream#buffer(name, buffer, filename[, contentType])

Add a buffer as a file to upload.

#### Arguments

- **name** String - Name of field
- **buffer** Buffer - The buffer to be uploaded
- **filename** String - The file name that tells the remote server
- ***contentType*** String - Optional. Content-Type (aka. MIME Type) of content (will be infered with `filename` if empty)

#### Returns

Form - form instance

### FormStream#stream(name, stream, filename[, contentType][, size])

Add a readable stream as a file to upload. Event 'error' will be emitted if an error occured.

#### Arguments

- **name** String - Name of field
- **stream** [stream.Readable](http://nodejs.org/api/stream.html#stream_class_stream_readable) - A readable stream to be piped
- **filename** String - The file name that tells the remote server
- ***contentType*** String - Optional. Content-Type (aka. MIME Type) of content (will be infered with `filename` if empty)
- ***size*** Number - Optional. Size of the stream (will not generate `Content-Length` header if not specified)

#### Returns

Form - form instance

### FormStream#headers([headers])

Get headers for the request.

#### Arguments

- ***headers*** Object - Additional headers

#### Example

```js
var headers = form.headers({
  'Authorization': 'Bearer kei2akc92jmznvnkeh09sknzdk',
  'Accept': 'application/vnd.github.v3.full+json'
});
```

#### Returns

Object - Headers to be sent.

### Event 'error'

Emitted if there was an error receiving data.

### Event 'data'

The 'data' event emits when a Buffer was used.

See [Node.js Documentation](http://nodejs.org/api/stream.html#stream_event_data) for more.

### Event 'end'

Emitted when the stream has received no more 'data' events will happen.

See [Node.js Documentation](http://nodejs.org/api/stream.html#stream_event_end) for more.

## License

(The MIT License)

Copyright (c) 2012 - 2014 fengmk2 &lt;fengmk2@gmail.com&gt;
Copyright(c) node-modules and other contributors.

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

1.2.0 / 2023-03-17
==================

**features**
  * [[`f268e86`](http://github.com/node-modules/formstream/commit/f268e86e6b55d5b981bb7daf9c8662bdcd820f5c)] - feat: Porting from new Buffer() to Buffer.from() (#22) (Matthew Speiser <<mattrspeiser@gmail.com>>)

**others**
  * [[`3c09f70`](http://github.com/node-modules/formstream/commit/3c09f700b2440b8f391650c939d91c2d5e4ae9b3)] - test: run on github action (#21) (fengmk2 <<fengmk2@gmail.com>>)
  * [[`2d5ed81`](http://github.com/node-modules/formstream/commit/2d5ed81e95b4fde270b5db2720fff0e19effd2dc)] - test: run test on node 12, 14, 16 (#20) (fengmk2 <<fengmk2@gmail.com>>)

1.1.1 / 2021-04-28
==================

**fixes**
  * [[`c96ba5b`](http://github.com/node-modules/formstream/commit/c96ba5bace0e96bf39770769e43d9de4271971d8)] - fix: upgrade mime to fix wrong mime type (#19) (shaozj <<shaozj@users.noreply.github.com>>)

**others**
  * [[`21918e5`](http://github.com/node-modules/formstream/commit/21918e5fc37a4cea5aae82659d39587f099e805b)] - chore: upgrade devDeps express to latest version (#15) (fengmk2 <<fengmk2@gmail.com>>)

1.1.0 / 2016-12-19
==================

  * deps: upgrade mime to latest version (#9)

1.0.0 / 2014-11-04
==================

 * fix(field): support stream.field(String, Number)
 * chore: fix links
 * chore: use npm scripts instead of Makefile
 * fix test case on node@0.8

0.0.8 / 2014-01-17 
==================

  * destroy source stream when formstream destroy()
  * add more test cases

0.0.7 / 2013-07-25 
==================

  * feature: always try to infer `Content-Length` (@xingrz)
  * doc improve

0.0.6 / 2013-07-15 
==================

  * added test cases for chaining call (@xingrz)
  * improved docs (@xingrz)
  * added chaining support (@xingrz)
  * api doc
  * fixed test causes
  * update dependencies version

0.0.5 / 2012-11-06 
==================

  * fixed stream error not catch bug

0.0.4 / 2012-11-06 
==================

  * fixed #2 support form.buffer()
  * add doc for setTotalStreamSize()

0.0.3 / 2012-11-06 
==================

  * support content-length use form.setTotalStreamSize(size)
  * update readme

0.0.2 / 2012-10-11 
==================

  * use buffer-concat support node < 0.8

0.0.1 / 2012-10-11 
==================

  * support multi streams.
  * add one stream support now.
  * Initial commit


test/fixtures/server.js
'use strict';

var multipart = require('connect-multiparty');
var express = require('express');
var app = express();

app.use(multipart());
app.use(function (req, res) {
  var files = {};
  for (var k in req.files) {
    var f = req.files[k];
    files[k] = {
      size: f.size || f.length,
      mime: f.type || f.mime,
      filename: f.name || f.filename,
      path: f.path
    };
  }
  res.send({
    url: req.url,
    method: req.method,
    headers: req.headers,
    body: req.body,
    files: files
  });
});

module.exports = app;


test/formstream.test.js
'use strict';

const Os = require('os');
Os.tmpDir = Os.tmpDir || Os.tmpdir;

var pedding = require('pedding');
var Stream = require('stream');
var http = require('http');
var fs = require('fs');
var path = require('path');
var should = require('should');
var urllib = require('urllib');
var formstream = require('../');

var root = path.join(__dirname, 'fixtures');
var app = require('./fixtures/server');

function cunterStream(name, count) {
  var s = new Stream();
  s.size = 0;
  var timer = setInterval(function () {
    var data = name + ' counter stream' + count + '\r\n';
    s.size += data.length;
    s.emit('data', data);
    count--;
    if (count <= 0) {
      clearInterval(timer);
      process.nextTick(function () {
        s.emit('end');
      });
    }
  }, 100);
  return s;
}

function post(port, url, form, callback) {
  var options = {
    port: port,
    host: '127.0.0.1',
    path: url,
    method: 'POST',
    headers: form.headers()
  };
  var done = function (err, data) {
    form.destroy();
    callback(err, data);
  };
  var req = http.request(options);
  req.on('response', function (res) {
    var chunks = [];
    res.on('data', function (chunk) {
      chunks.push(chunk);
    });
    res.on('end', function () {
      var data = Buffer.concat(chunks);
      var err = null;
      try {
        data = JSON.parse(data);
      } catch (e) {
        err = e;
        err.data = data.toString();
      }
      done(err, data);
    });
  });
  form.pipe(req);
  req.on('error', done);
  form.on('error', req.emit.bind(req, 'error'));
}

describe('formstream.test.js', function () {
  var port;
  before(function (done) {
    app = app.listen(0, function () {
      port = app.address().port;
      done();
    });
  });

  it('should post fields only with content-length', function (done) {
    done = pedding(2, done);
    var form = formstream();
    form.field('foo', 'bar');
    form.field('name', '中文名字');
    form.field('pwd', '哈哈pwd');
    form.on('destroy', done);
    post(port, '/post', form, function (err, data) {
      should.not.exist(err);
      data.body.should.eql({
        foo: 'bar',
        name: '中文名字',
        pwd: '哈哈pwd'
      });
      data.headers.should.have.property('content-length', String(form._contentLength));
      data.headers.should.have.property('content-type')
        .with.equal('multipart/form-data; boundary=' + form._boundary);
      data.files.should.eql({});
      done(err);
    });
  });

  it.skip('should upload a stream without size, use "Transfer-Encoding: chunked"', function (done) {
    var ChunkedStream = require('../../chunked');
    var form = formstream();
    form.stream('file', fs.createReadStream(path.join(__dirname, 'fixtures', 'foo.txt')), 'foo.txt');
    form.field('name', '哈哈');
    var headers = form.headers();
    headers['Transfer-Encoding'] = 'chunked';
    var chunked = new ChunkedStream();
    form.pipe(chunked);

    var options = {
      method: 'post',
      headers: headers,
      // dataType: 'json',
      stream: chunked,
    };
    var url = 'http://127.0.0.1:' + port;
    // url = 'http://cgi-lib.berkeley.edu/ex/fup.cgi';

    urllib.request(url, options, function (err, data, res) {
      should.not.exist(err);
      // console.log(data.toString());
      res.should.status(200);
      done();
    });
    // chunked.pipe(process.stdout);
  });

  it('should post fields and file', function (done) {
    done = pedding(2, done);
    var now = Date.now();
    var form = formstream();
    form.field('foo', 'bar');
    form.field('name', '中文名字');
    form.field('pwd', '哈哈pwd');
    form.field('now', now);
    form.file('file', __filename);
    form.on('destroy', done);
    post(port, '/post', form, function (err, data) {
      data.body.should.eql({
        foo: 'bar',
        name: '中文名字',
        pwd: '哈哈pwd',
        now: String(now)
      });
      data.headers.should.not.have.property('content-length');
      data.headers.should.have.property('content-type')
        .with.equal('multipart/form-data; boundary=' + form._boundary);
      var files = data.files;
      files.should.have.property('file');
      files.file.filename.should.equal('formstream.test.js');
      files.file.size.should.equal(fs.statSync(__filename).size);
      files.file.mime.should.equal('application/javascript');
      done(err);
    });
  });

  it('should post fields and file with content-length', function (done) {
    fs.stat(__filename, function (err, stat) {
      var form = formstream();
      form.field('foo', 'bar');
      form.field('name', '中文名字');
      form.field('pwd', '哈哈pwd');
      form.file('file', __filename);
      form.setTotalStreamSize(stat.size);
      post(port, '/post', form, function (err, data) {
        data.body.should.eql({
          foo: 'bar',
          name: '中文名字',
          pwd: '哈哈pwd'
        });
        data.headers.should.have.property('content-length').with.equal(String(form._contentLength));
        data.headers.should.have.property('content-type')
          .with.equal('multipart/form-data; boundary=' + form._boundary);
        var files = data.files;
        files.should.have.property('file');
        files.file.filename.should.equal('formstream.test.js');
        files.file.size.should.equal(fs.statSync(__filename).size);
        files.file.mime.should.equal('application/javascript');
        done(err);
      });
    });
  });

  it('should post fields and file with wrong stream size will return error', function (done) {
    done = pedding(2, done);
    var form = formstream();
    form.field('foo', 'bar');
    form.field('name', '中文名字');
    form.field('pwd', '哈哈pwd');
    form.file('file', __filename);
    form.setTotalStreamSize(100);
    form.on('destroy', done);
    post(port, '/post', form, function (err) {
      should.exist(err);
      done();
    });
  });

  it('should post fields and file with wrong size will return error', function (done) {
    done = pedding(2, done);
    var form = formstream();
    form.field('foo', 'bar');
    form.field('name', '中文名字');
    form.field('pwd', '哈哈pwd');
    form.file('file', __filename, 100);
    form.on('destroy', done);
    post(port, '/post', form, function (err) {
      should.exist(err);
      done();
    });
  });

  if (process.version.indexOf('v0.8.') !== 0) {
    // node 0.8, createSteram not exists file will throw error
    it('should post not exist file return error ENOENT', function (done) {
      var form = formstream();
      form.field('foo', 'bar');
      form.field('name', '中文名字');
      form.field('pwd', '哈哈pwd');
      form.file('file', __filename + 'notexists');
      form.setTotalStreamSize(100);
      post(port, '/post', form, function (err) {
        should.exist(err);
        err.message.should.containEql('formstream/test/formstream.test.jsnotexists');
        err.message.should.containEql('ENOENT');
        done();
      });
    });
  }

  it('should post fields and stream', function (done) {
    var form = formstream();
    var s1 = cunterStream('no1', 5);
    form.stream('stream1', s1, 'stream1中文名.txt', 'text/html');
    var s2 = cunterStream('no2', 3);
    form.stream('stream2', s2, 'stream2.png');
    form.field('foo', 'bar');
    form.field('name', '中文名字');
    form.field('pwd', '哈哈pwd');
    form.file('file', __filename);

    post(port, '/post', form, function (err, data) {
      data.body.should.eql({
        foo: 'bar',
        name: '中文名字',
        pwd: '哈哈pwd'
      });
      data.headers.should.have.property('content-type')
        .with.match(/multipart\/form-data; boundary=--------------------------\d{24}/);
      var files = data.files;
      files.should.have.keys('stream1', 'stream2', 'file');
      var stream1 = files.stream1;
      var stream2 = files.stream2;
      stream1.filename.should.equal('stream1中文名.txt');
      stream1.size.should.equal(stream1.size);
      stream1.mime.should.equal('text/html');

      stream2.filename.should.equal('stream2.png');
      stream2.size.should.equal(stream2.size);
      stream2.mime.should.equal('image/png');

      files.should.have.property('file');
      files.file.filename.should.equal('formstream.test.js');
      files.file.size.should.equal(fs.statSync(__filename).size);
      files.file.mime.should.equal('application/javascript');
      done(err);
    });
  });

  it('should post fields, 2 file', function (done) {
    var form = formstream();
    form.field('foo', 'bar');
    form.field('name', '中文名字');
    form.field('pwd', '哈哈pwd');
    form.file('file', __filename);
    form.file('logo', path.join(root, 'logo.png'));
    post(port, '/post', form, function (err, data) {
      data.body.should.eql({
        foo: 'bar',
        name: '中文名字',
        pwd: '哈哈pwd'
      });
      data.headers.should.have.property('content-type')
        .with.match(/^multipart\/form-data; boundary=--------------------------\d{24}$/);
      var files = data.files;
      files.should.have.keys('file', 'logo');
      files.file.filename.should.equal('formstream.test.js');
      files.file.size.should.equal(fs.statSync(__filename).size);
      files.file.mime.should.equal('application/javascript');
      done(err);
    });
  });

  it('should post fields, 2 file with content-length', function (done) {
    var size = 0;
    var ready = function () {
      var form = formstream();
      form.field('foo', 'bar');
      form.field('name', '中文名字');
      form.field('pwd', '哈哈pwd');
      form.file('file', __filename);
      form.file('logo', path.join(root, 'logo.png'));
      form.setTotalStreamSize(size);
      post(port, '/post', form, function (err, data) {
        data.body.should.eql({
          foo: 'bar',
          name: '中文名字',
          pwd: '哈哈pwd'
        });
        data.headers.should.have.property('content-length').with.equal(form._contentLength + '');
        data.headers.should.have.property('content-type')
          .with.match(/^multipart\/form-data; boundary=--------------------------\d{24}$/);
        var files = data.files;
        files.should.have.keys('file', 'logo');
        files.file.filename.should.equal('formstream.test.js');
        files.file.size.should.equal(fs.statSync(__filename).size);
        files.file.mime.should.equal('application/javascript');
        done(err);
      });
    };

    ready = pedding(2, ready);
    fs.stat(__filename, function (err, stat) {
      size += stat.size;
      ready();
    });
    fs.stat(path.join(root, 'logo.png'), function (err, stat) {
      size += stat.size;
      ready();
    });
  });

  describe('buffer()', function () {
    it('should post file content buffer', function (done) {
      var form = formstream();
      form.field('foo', 'bar');
      form.field('name', '中文名字');
      form.field('pwd', '哈哈pwd');
      var buffer = Buffer.from('foo content');
      form.buffer('file', buffer, 'foo.txt');
      var bar = Buffer.from('bar content中文');
      form.buffer('bar', bar, 'bar.js');
      var logopath = path.join(root, 'logo.png');
      form.file('logo', logopath);
      post(port, '/post', form, function (err, data) {
        data.body.should.eql({
          foo: 'bar',
          name: '中文名字',
          pwd: '哈哈pwd'
        });
        data.headers.should.have.property('content-type')
          .with.equal('multipart/form-data; boundary=' + form._boundary);
        var files = data.files;
        files.should.have.keys('file', 'logo', 'bar');
        files.file.filename.should.equal('foo.txt');
        files.file.size.should.equal(buffer.length);
        files.file.mime.should.equal('text/plain');

        files.bar.filename.should.equal('bar.js');
        files.bar.size.should.equal(bar.length);
        files.bar.mime.should.equal('application/javascript');
        fs.readFileSync(files.bar.path, 'utf8').should.equal('bar content中文');

        files.logo.filename.should.equal('logo.png');
        files.logo.size.should.equal(fs.statSync(logopath).size);
        files.logo.mime.should.equal('image/png');
        done(err);
      });
    });

    it('should post file content buffer with content-length', function (done) {
      var form = formstream();
      form.field('foo', 'bar');
      form.field('name', '中文名字');
      form.field('pwd', '哈哈pwd');
      var buffer = Buffer.from('file content');
      form.buffer('file', buffer, 'foo.txt');
      var logopath = path.join(root, 'logo.png');
      form.file('logo', logopath);
      form.setTotalStreamSize(fs.statSync(logopath).size);
      post(port, '/post', form, function (err, data) {
        data.body.should.eql({
          foo: 'bar',
          name: '中文名字',
          pwd: '哈哈pwd'
        });
        data.headers.should.have.property('content-type')
          .with.equal('multipart/form-data; boundary=' + form._boundary);
        var files = data.files;
        files.should.have.keys('file', 'logo');
        files.file.filename.should.equal('foo.txt');
        files.file.size.should.equal(buffer.length);
        files.file.mime.should.equal('text/plain');

        files.logo.filename.should.equal('logo.png');
        files.logo.size.should.equal(fs.statSync(logopath).size);
        files.logo.mime.should.equal('image/png');
        done(err);
      });
    });
  });

  describe('headers()', function () {
    it('should get headers contains correct content-length', function () {
      var form = formstream();
      form.field('foo', 'bar');
      var headers = form.headers({ 'X-Test': 'hello' });
      headers.should.have.keys('Content-Type', 'Content-Length', 'X-Test');
      headers['Content-Type'].should.match(/^multipart\/form-data; boundary=--------------------------\d{24}$/);
      headers['X-Test'].should.equal('hello');
      headers['Content-Length'].should.equal('161');
    });

    it('should trust .setTotalStreamSize() only if has any stream/file without size specified', function () {
      var headers1 = formstream()
        .field('field', 'plain')
        .file('file', path.join(root, 'logo.png'), 'file')
        .buffer('buffer', Buffer.alloc(20), 'buffer')
        .stream('stream', cunterStream('stream', 5), 'stream')
        .setTotalStreamSize(10)
        .headers();

      var headers2 = formstream()
        .field('field', 'plain')
        .file('file', path.join(root, 'logo.png'), 'file', 10)
        .buffer('buffer', Buffer.alloc(20), 'buffer')
        .stream('stream', cunterStream('stream', 5), 'stream', 30)
        .headers();

      headers1.should.have.property('Content-Length');
      headers2.should.have.property('Content-Length');

      var length1 = parseInt(headers1['Content-Length']);
      var length2 = parseInt(headers2['Content-Length']);

      length1.should.equal(length2 - 30);
    });
  });

  describe('chaining', function () {
    it('should do chaining calls with .field()', function () {
      var form = formstream();
      form.field('foo', 'bar').should.equal(form);
    });

    it('should do chaining calls with .file()', function () {
      var form = formstream();
      form.file('foo', path.join(root, 'logo.png'), 'bar').should.equal(form);
    });

    it('should do chaining calls with .buffer()', function () {
      var form = formstream();
      form.buffer('foo', Buffer.from('foo content'), 'bar').should.equal(form);
    });

    it('should do chaining calls with .stream()', function () {
      var form = formstream();
      form.stream('foo', cunterStream('stream', 5), 'bar').should.equal(form);
    });

    it('should do chaining calls with .setTotalStreamSize()', function () {
      var form = formstream();
      form.setTotalStreamSize(10).should.equal(form);
    });
  });

  it('should get right mime with emf file', function (done) {
    done = pedding(2, done);
    var form = formstream();
    form.field('foo', 'bar');
    form.file('file', path.join(root, 'test.emf'));
    form.on('destroy', done);
    post(port, '/post', form, function (err, data) {
      data.headers.should.not.have.property('content-length');
      data.headers.should.have.property('content-type')
        .with.equal('multipart/form-data; boundary=' + form._boundary);
      var files = data.files;
      files.should.have.property('file');
      files.file.filename.should.equal('test.emf');
      files.file.mime.should.equal('image/emf');
      done(err);
    });
  });

});


example/upload.js
'use strict';

var http = require('http');
var fs = require('fs');
var path = require('path');
var formstream = require('../');

var imagepath = path.join(path.dirname(__dirname), 'logo.png');

var form = formstream();
// form.file('file', filepath, filename);
form.stream('file', fs.createReadStream(imagepath), 'logo.png');
form.field('foo', 'hello world');

var req = http.request({
  method: 'POST',
  host: 'upload.cnodejs.net',
  // host: '127.0.0.1',
  // port: 8081,
  path: '/store',
  headers: form.headers()
});
req.on('response', function (res) {
  console.log(res.statusCode, res.headers);
  var chunks = [];
  res.on('data', function (chunk) {
    chunks.push(chunk);
  }).on('end', function () {
    console.log('%s', Buffer.concat(chunks).toString());
    console.log('upload success.');
    process.exit(0);
  });
});

var size = 0;
form.on('data', function (data) {
  size += data.length;
  console.log('uploading... %d bytes', size);
  req.write(data);
}).on('end', function () {
  console.log('uploaded %d bytes', size);
  req.end();
});


lib/formstream.js
/**
 * Data format:
 *

--FormStreamBoundary1349886663601\r\n
Content-Disposition: form-data; name="foo"\r\n
\r\n
<FIELD-CONTENT>\r\n
--FormStreamBoundary1349886663601\r\n
Content-Disposition: form-data; name="file"; filename="formstream.test.js"\r\n
Content-Type: application/javascript\r\n
\r\n
<FILE-CONTENT>\r\n
--FormStreamBoundary1349886663601\r\n
Content-Disposition: form-data; name="pic"; filename="fawave.png"\r\n
Content-Type: image/png\r\n
\r\n
<IMAGE-CONTENT>\r\n
--FormStreamBoundary1349886663601--

 *
 */

'use strict';

var Stream = require('stream');
var parseStream = require('pause-stream');
var util = require('util');
var mime = require('mime');
var path = require('path');
var fs = require('fs');
var destroy = require('destroy');

var PADDING = '--';
var NEW_LINE = '\r\n';
var NEW_LINE_BUFFER =  Buffer.from(NEW_LINE);

function FormStream() {
  if (!(this instanceof FormStream)) {
    return new FormStream();
  }

  FormStream.super_.call(this);

  this._boundary = this._generateBoundary();
  this._streams = [];
  this._buffers = [];
  this._endData = Buffer.from(PADDING + this._boundary + PADDING + NEW_LINE);
  this._contentLength = 0;
  this._isAllStreamSizeKnown = true;
  this._knownStreamSize = 0;
}

util.inherits(FormStream, Stream);
module.exports = FormStream;

FormStream.prototype._generateBoundary = function() {
  // https://github.com/felixge/node-form-data/blob/master/lib/form_data.js#L162
  // This generates a 50 character boundary similar to those used by Firefox.
  // They are optimized for boyer-moore parsing.
  var boundary = '--------------------------';
  for (var i = 0; i < 24; i++) {
    boundary += Math.floor(Math.random() * 10).toString(16);
  }

  return boundary;
};

FormStream.prototype.setTotalStreamSize = function (size) {
  // this method should not make any sense if the length of each stream is known.
  if (this._isAllStreamSizeKnown) {
    return this;
  }

  size = size || 0;

  for (var i = 0; i < this._streams.length; i++) {
    size += this._streams[i][0].length;
    size += NEW_LINE_BUFFER.length; // stream field end pedding size
  }

  this._knownStreamSize = size;
  this._isAllStreamSizeKnown = true;

  return this;
};

FormStream.prototype.headers = function (options) {
  var headers = {
    'Content-Type': 'multipart/form-data; boundary=' + this._boundary
  };

  // calculate total stream size
  this._contentLength += this._knownStreamSize;

  // calculate length of end padding
  this._contentLength += this._endData.length;

  if (this._isAllStreamSizeKnown) {
    headers['Content-Length'] = String(this._contentLength);
  }

  if (options) {
    for (var k in options) {
      headers[k] = options[k];
    }
  }

  return headers;
};

FormStream.prototype.file = function (name, filepath, filename, filesize) {
  var mimeType = mime.getType(filepath);

  if (typeof filename === 'number' && !filesize) {
    filesize = filename;
    filename = path.basename(filepath);
  } else if (!filename) {
    filename = path.basename(filepath);
  }

  var stream = fs.createReadStream(filepath);

  return this.stream(name, stream, filename, mimeType, filesize);
};

/**
 * Add a form field
 * @param  {String} name field name
 * @param  {String|Buffer} value field value
 * @return {this}
 */
FormStream.prototype.field = function (name, value) {
  if (!Buffer.isBuffer(value)) {
    // field(String, Number)
    // https://github.com/qiniu/nodejs-sdk/issues/123
    if (typeof value === 'number') {
      value = String(value);
    }
    value = Buffer.from(value);
  }
  return this.buffer(name, value);
};

FormStream.prototype.stream = function (name, stream, filename, mimeType, size) {
  if (typeof mimeType === 'number' && !size) {
    size = mimeType;
    mimeType = mime.getType(filename);
  } else if (!mimeType) {
    mimeType = mime.getType(filename);
  }

  stream.once('error', this.emit.bind(this, 'error'));
  // if form stream destroy, also destroy the source stream
  this.once('destroy', function () {
    destroy(stream);
  });

  var leading = this._leading({ name: name, filename: filename }, mimeType);

  var ps = parseStream().pause();
  stream.pipe(ps);

  this._streams.push([leading, ps]);

  // if the size of this stream is known, plus the total content-length;
  // otherwise, content-length is unknown.
  if (typeof size === 'number') {
    this._knownStreamSize += leading.length;
    this._knownStreamSize += size;
    this._knownStreamSize += NEW_LINE_BUFFER.length;
  } else {
    this._isAllStreamSizeKnown = false;
  }

  process.nextTick(this.resume.bind(this));

  return this;
};

FormStream.prototype.buffer = function (name, buffer, filename, mimeType) {
  if (filename && !mimeType) {
    mimeType = mime.getType(filename);
  }

  var disposition = { name: name };
  if (filename) {
    disposition.filename = filename;
  }

  var leading = this._leading(disposition, mimeType);

  this._buffers.push([leading, buffer]);

  // plus buffer length to total content-length
  this._contentLength += leading.length;
  this._contentLength += buffer.length;
  this._contentLength += NEW_LINE_BUFFER.length;

  process.nextTick(this.resume.bind(this));

  return this;
};

FormStream.prototype._leading = function (disposition, type) {
  var leading = [PADDING + this._boundary];

  var disps = [];

  if (disposition) {
    for (var k in disposition) {
      disps.push(k + '="' + disposition[k] + '"');
    }
  }

  leading.push('Content-Disposition: form-data; ' + disps.join('; '));

  if (type) {
    leading.push('Content-Type: ' + type);
  }

  leading.push('');
  leading.push('');

  return Buffer.from(leading.join(NEW_LINE));
};

FormStream.prototype._emitBuffers = function () {
  if (!this._buffers.length) {
    return;
  }

  for (var i = 0; i < this._buffers.length; i++) {
    var item = this._buffers[i];
    this.emit('data', item[0]); // part leading
    this.emit('data', item[1]); // part content
    this.emit('data', NEW_LINE_BUFFER);
  }

  this._buffers = [];
};

FormStream.prototype._emitStream = function (item) {
  var self = this;
  // item: [ fieldData, stream ]
  self.emit('data', item[0]);

  var stream = item[1];
  stream.on('data', function (data) {
    self.emit('data', data);
  });
  stream.on('end', function () {
    self.emit('data', NEW_LINE_BUFFER);
    return process.nextTick(self.drain.bind(self));
  });
  stream.resume();
};

FormStream.prototype._emitEnd = function () {
  // ending format:
  //
  // --{boundary}--\r\n
  this.emit('data', this._endData);
  this.emit('end');
};

FormStream.prototype.drain = function () {
  this._emitBuffers();

  var item = this._streams.shift();
  if (item) {
    this._emitStream(item);
  } else {
    this._emitEnd();
  }

  return this;
};

FormStream.prototype.resume = function () {
  this.paused = false;

  if (!this._draining) {
    this._draining = true;
    this.drain();
  }

  return this;
};

FormStream.prototype.close = FormStream.prototype.destroy = function () {
  this.emit('destroy');
};
